dask.bag.from_s3

cowlicks opened the pull request:
Ready for review.

~~Not the best API right now.~~

~~This is not working. Something is wrong with the dask I am making [here](https://github.com/ContinuumIO/dask/pull/245/files#diff-8fd8f45b66540dedbca6875ddd2f3579R771). When I call `compute` on the bag it just returns the dask.~~

cowlicks commented:
Closes #241

cowlicks commented:
@shoyer how would you change the API?

cowlicks commented:
Any ideas how this should be tested with Travis-CI?

mrocklin commented:
> Any ideas how this should be tested with Travis-CI? 

Throw up a public dataset on s3?  I think that odo uses `s3://nyqpug/tips.csv`.

cowlicks commented:
Note that there is a pickling error in this when `anon=True` and we use an async scheduler.

I think part of boto is not threadsafe when `anon=True`. But it is when we supply credentials? Since it is threadsafe sometimes it might be an easy fix on Boto's end.

```python
In [1]: import dask.bag as db

In [2]: x = db.from_s3('mtg-sets', '10E-x.json', anon=True)

In [3]: x.compute()
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-3-f4c69c9c1276> in <module>()
----> 1 x.compute()

/home/blake/github/dask/dask/bag/core.pyc in compute(self, **kwargs)
    571     def compute(self, **kwargs):
    572         """ Force evaluation of bag """
--> 573         results = get(self.dask, self._keys(), **kwargs)
    574         if isinstance(results[0], Iterable) and not isinstance(results[0], str):
    575             results = toolz.concat(results)

/home/blake/github/dask/dask/bag/core.pyc in get(dsk, keys, get, **kwargs)
     83     dsk2 = optimize(dsk, keys)
     84 
---> 85     return get(dsk2, keys, **kwargs)
     86 _get = get
     87 

/home/blake/github/dask/dask/multiprocessing.pyc in get(dsk, keys, optimizations, num_workers, func_loads, func_dumps)
     49         # Run
     50         result = get_async(apply_async, len(pool._pool), dsk2, keys,
---> 51                            queue=queue)
     52     finally:
     53         if cleanup:

/home/blake/github/dask/dask/async.pyc in get_async(apply_async, num_workers, dsk, result, cache, debug_counts, queue, raise_on_exception, **kwargs)
    438     # Seed initial tasks into the thread pool
    439     while state['ready'] and len(state['running']) < num_workers:
--> 440         fire_task()
    441 
    442     # Main loop, wait on tasks to finish, insert new ones

/home/blake/github/dask/dask/async.pyc in fire_task()
    434         # Submit
    435         apply_async(execute_task, args=[key, dsk[key], data, queue,
--> 436                                         raise_on_exception])
    437 
    438     # Seed initial tasks into the thread pool

/home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/toolz/functoolz.pyc in __call__(self, *args, **kwargs)
    217     def __call__(self, *args, **kwargs):
    218         try:
--> 219             return self._partial(*args, **kwargs)
    220         except TypeError:
    221             # If there was a genuine TypeError

/home/blake/github/dask/dask/multiprocessing.pyc in dill_apply_async(apply_async, func, args, kwds, func_loads, func_dumps)
     68     dumps = func_dumps or _globals.get('func_dumps') or dill.dumps
     69     sfunc = dumps(func)
---> 70     sargs = dumps(args)
     71     skwds = dumps(kwds)
     72     return apply_async(curry(apply_func, loads=func_loads),

/home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in dumps(obj, protocol, byref, fmode)
    190     """pickle an object to a string"""
    191     file = StringIO()
--> 192     dump(obj, file, protocol, byref, fmode)#, strictio)
    193     return file.getvalue()
    194 

/home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in dump(obj, file, protocol, byref, fmode)
    180             return
    181     # end hack
--> 182     pik.dump(obj)
    183     # return to saved settings
    184     pik._byref = _byref

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in dump(self, obj)
    222         if self.proto >= 2:
    223             self.write(PROTO + chr(self.proto))
--> 224         self.save(obj)
    225         self.write(STOP)
    226 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_list(self, obj)
    598 
    599         self.memoize(obj)
--> 600         self._batch_appends(iter(obj))
    601 
    602     dispatch[ListType] = save_list

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in _batch_appends(self, items)
    631                 write(MARK)
    632                 for x in tmp:
--> 633                     save(x)
    634                 write(APPENDS)
    635             elif n:

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_tuple(self, obj)
    546         if n <= 3 and proto >= 2:
    547             for element in obj:
--> 548                 save(element)
    549             # Subtle.  Same as in the big comment below.
    550             if id(obj) in memo:

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_tuple(self, obj)
    546         if n <= 3 and proto >= 2:
    547             for element in obj:
--> 548                 save(element)
    549             # Subtle.  Same as in the big comment below.
    550             if id(obj) in memo:

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288 

/home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in save_instancemethod0(pickler, obj)
    792     else:
    793         pickler.save_reduce(MethodType, (obj.im_func, obj.im_self,
--> 794                                          obj.im_class), obj=obj)
    795     return
    796 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems, obj)
    399         else:
    400             save(func)
--> 401             save(args)
    402             write(REDUCE)
    403 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_tuple(self, obj)
    546         if n <= 3 and proto >= 2:
    547             for element in obj:
--> 548                 save(element)
    549             # Subtle.  Same as in the big comment below.
    550             if id(obj) in memo:

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    329 
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332 
    333     def persistent_id(self, obj):

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems, obj)
    417 
    418         if state is not None:
--> 419             save(state)
    420             write(BUILD)
    421 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288 

/home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in save_module_dict(pickler, obj)
    656             # we only care about session the first pass thru
    657             pickler._session = False
--> 658         StockPickler.save_dict(pickler, obj)
    659     return
    660 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_dict(self, obj)
    647 
    648         self.memoize(obj)
--> 649         self._batch_setitems(obj.iteritems())
    650 
    651     dispatch[DictionaryType] = save_dict

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in _batch_setitems(self, items)
    679                 for k, v in tmp:
    680                     save(k)
--> 681                     save(v)
    682                 write(SETITEMS)
    683             elif n:

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    329 
    330         # Save the reduce() output and finally memoize the object
--> 331         self.save_reduce(obj=obj, *rv)
    332 
    333     def persistent_id(self, obj):

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems, obj)
    417 
    418         if state is not None:
--> 419             save(state)
    420             write(BUILD)
    421 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    284         f = self.dispatch.get(t)
    285         if f:
--> 286             f(self, obj) # Call unbound method with explicit self
    287             return
    288 

/home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in save_module_dict(pickler, obj)
    656             # we only care about session the first pass thru
    657             pickler._session = False
--> 658         StockPickler.save_dict(pickler, obj)
    659     return
    660 

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_dict(self, obj)
    647 
    648         self.memoize(obj)
--> 649         self._batch_setitems(obj.iteritems())
    650 
    651     dispatch[DictionaryType] = save_dict

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in _batch_setitems(self, items)
    679                 for k, v in tmp:
    680                     save(k)
--> 681                     save(v)
    682                 write(SETITEMS)
    683             elif n:

/home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
    304             reduce = getattr(obj, "__reduce_ex__", None)
    305             if reduce:
--> 306                 rv = reduce(self.proto)
    307             else:
    308                 reduce = getattr(obj, "__reduce__", None)

/home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/boto/auth.pyc in __getstate__(self)
    102     def __getstate__(self):
    103         pickled_dict = copy.copy(self.__dict__)
--> 104         del pickled_dict['_hmac']
    105         del pickled_dict['_hmac_256']
    106         return pickled_dict

KeyError: '_hmac'

mrocklin commented:
Perhaps we construct the boto connection every time we do a single read?
Serializing something as stateful as a connection might not work well.  I
wouldn't expect boto devs to jump to solve this problem.

If it takes a long time to initialize the connection on each blocked read
then we run into a problem.  Caching might resolve that if it's an issue
(although this also might not be an issue).

On Thu, May 28, 2015 at 9:40 PM, Blake Griffith <notifications@github.com>
wrote:

> Note that there is a pickling error in this when anon=True and we use an
> async scheduler.
>
> I think part of boto is not threadsafe when anon=True. But it is when we
> supply credentials? Since it is threadsafe sometimes it might be an easy
> fix on Boto's end.
>
> In [1]: import dask.bag as db
>
> In [2]: x = db.from_s3('mtg-sets', '10E-x.json', anon=True)
>
> In [3]: x.compute()---------------------------------------------------------------------------KeyError                                  Traceback (most recent call last)<ipython-input-3-f4c69c9c1276> in <module>()----> 1 x.compute()
> /home/blake/github/dask/dask/bag/core.pyc in compute(self, **kwargs)
>     571     def compute(self, **kwargs):
>     572         """ Force evaluation of bag """--> 573         results = get(self.dask, self._keys(), **kwargs)
>     574         if isinstance(results[0], Iterable) and not isinstance(results[0], str):
>     575             results = toolz.concat(results)
> /home/blake/github/dask/dask/bag/core.pyc in get(dsk, keys, get, **kwargs)
>      83     dsk2 = optimize(dsk, keys)
>      84 ---> 85     return get(dsk2, keys, **kwargs)
>      86 _get = get
>      87
> /home/blake/github/dask/dask/multiprocessing.pyc in get(dsk, keys, optimizations, num_workers, func_loads, func_dumps)
>      49         # Run
>      50         result = get_async(apply_async, len(pool._pool), dsk2, keys,---> 51                            queue=queue)
>      52     finally:
>      53         if cleanup:
> /home/blake/github/dask/dask/async.pyc in get_async(apply_async, num_workers, dsk, result, cache, debug_counts, queue, raise_on_exception, **kwargs)
>     438     # Seed initial tasks into the thread pool
>     439     while state['ready'] and len(state['running']) < num_workers:--> 440         fire_task()
>     441
>     442     # Main loop, wait on tasks to finish, insert new ones
> /home/blake/github/dask/dask/async.pyc in fire_task()
>     434         # Submit
>     435         apply_async(execute_task, args=[key, dsk[key], data, queue,--> 436                                         raise_on_exception])
>     437
>     438     # Seed initial tasks into the thread pool
> /home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/toolz/functoolz.pyc in __call__(self, *args, **kwargs)
>     217     def __call__(self, *args, **kwargs):
>     218         try:--> 219             return self._partial(*args, **kwargs)
>     220         except TypeError:
>     221             # If there was a genuine TypeError
> /home/blake/github/dask/dask/multiprocessing.pyc in dill_apply_async(apply_async, func, args, kwds, func_loads, func_dumps)
>      68     dumps = func_dumps or _globals.get('func_dumps') or dill.dumps
>      69     sfunc = dumps(func)---> 70     sargs = dumps(args)
>      71     skwds = dumps(kwds)
>      72     return apply_async(curry(apply_func, loads=func_loads),
> /home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in dumps(obj, protocol, byref, fmode)
>     190     """pickle an object to a string"""
>     191     file = StringIO()--> 192     dump(obj, file, protocol, byref, fmode)#, strictio)
>     193     return file.getvalue()
>     194
> /home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in dump(obj, file, protocol, byref, fmode)
>     180             return
>     181     # end hack--> 182     pik.dump(obj)
>     183     # return to saved settings
>     184     pik._byref = _byref
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in dump(self, obj)
>     222         if self.proto >= 2:
>     223             self.write(PROTO + chr(self.proto))--> 224         self.save(obj)
>     225         self.write(STOP)
>     226
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     284         f = self.dispatch.get(t)
>     285         if f:--> 286             f(self, obj) # Call unbound method with explicit self
>     287             return
>     288
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_list(self, obj)
>     598
>     599         self.memoize(obj)--> 600         self._batch_appends(iter(obj))
>     601
>     602     dispatch[ListType] = save_list
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in _batch_appends(self, items)
>     631                 write(MARK)
>     632                 for x in tmp:--> 633                     save(x)
>     634                 write(APPENDS)
>     635             elif n:
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     284         f = self.dispatch.get(t)
>     285         if f:--> 286             f(self, obj) # Call unbound method with explicit self
>     287             return
>     288
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_tuple(self, obj)
>     546         if n <= 3 and proto >= 2:
>     547             for element in obj:--> 548                 save(element)
>     549             # Subtle.  Same as in the big comment below.
>     550             if id(obj) in memo:
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     284         f = self.dispatch.get(t)
>     285         if f:--> 286             f(self, obj) # Call unbound method with explicit self
>     287             return
>     288
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_tuple(self, obj)
>     546         if n <= 3 and proto >= 2:
>     547             for element in obj:--> 548                 save(element)
>     549             # Subtle.  Same as in the big comment below.
>     550             if id(obj) in memo:
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     284         f = self.dispatch.get(t)
>     285         if f:--> 286             f(self, obj) # Call unbound method with explicit self
>     287             return
>     288
> /home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in save_instancemethod0(pickler, obj)
>     792     else:
>     793         pickler.save_reduce(MethodType, (obj.im_func, obj.im_self,--> 794                                          obj.im_class), obj=obj)
>     795     return
>     796
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems, obj)
>     399         else:
>     400             save(func)--> 401             save(args)
>     402             write(REDUCE)
>     403
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     284         f = self.dispatch.get(t)
>     285         if f:--> 286             f(self, obj) # Call unbound method with explicit self
>     287             return
>     288
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_tuple(self, obj)
>     546         if n <= 3 and proto >= 2:
>     547             for element in obj:--> 548                 save(element)
>     549             # Subtle.  Same as in the big comment below.
>     550             if id(obj) in memo:
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     329
>     330         # Save the reduce() output and finally memoize the object--> 331         self.save_reduce(obj=obj, *rv)
>     332
>     333     def persistent_id(self, obj):
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems, obj)
>     417
>     418         if state is not None:--> 419             save(state)
>     420             write(BUILD)
>     421
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     284         f = self.dispatch.get(t)
>     285         if f:--> 286             f(self, obj) # Call unbound method with explicit self
>     287             return
>     288
> /home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in save_module_dict(pickler, obj)
>     656             # we only care about session the first pass thru
>     657             pickler._session = False--> 658         StockPickler.save_dict(pickler, obj)
>     659     return
>     660
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_dict(self, obj)
>     647
>     648         self.memoize(obj)--> 649         self._batch_setitems(obj.iteritems())
>     650
>     651     dispatch[DictionaryType] = save_dict
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in _batch_setitems(self, items)
>     679                 for k, v in tmp:
>     680                     save(k)--> 681                     save(v)
>     682                 write(SETITEMS)
>     683             elif n:
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     329
>     330         # Save the reduce() output and finally memoize the object--> 331         self.save_reduce(obj=obj, *rv)
>     332
>     333     def persistent_id(self, obj):
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_reduce(self, func, args, state, listitems, dictitems, obj)
>     417
>     418         if state is not None:--> 419             save(state)
>     420             write(BUILD)
>     421
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     284         f = self.dispatch.get(t)
>     285         if f:--> 286             f(self, obj) # Call unbound method with explicit self
>     287             return
>     288
> /home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/dill/dill.pyc in save_module_dict(pickler, obj)
>     656             # we only care about session the first pass thru
>     657             pickler._session = False--> 658         StockPickler.save_dict(pickler, obj)
>     659     return
>     660
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save_dict(self, obj)
>     647
>     648         self.memoize(obj)--> 649         self._batch_setitems(obj.iteritems())
>     650
>     651     dispatch[DictionaryType] = save_dict
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in _batch_setitems(self, items)
>     679                 for k, v in tmp:
>     680                     save(k)--> 681                     save(v)
>     682                 write(SETITEMS)
>     683             elif n:
> /home/blake/.miniconda/envs/dask/lib/python2.7/pickle.pyc in save(self, obj)
>     304             reduce = getattr(obj, "__reduce_ex__", None)
>     305             if reduce:--> 306                 rv = reduce(self.proto)
>     307             else:
>     308                 reduce = getattr(obj, "__reduce__", None)
> /home/blake/.miniconda/envs/dask/lib/python2.7/site-packages/boto/auth.pyc in __getstate__(self)
>     102     def __getstate__(self):
>     103         pickled_dict = copy.copy(self.__dict__)--> 104         del pickled_dict['_hmac']
>     105         del pickled_dict['_hmac_256']
>     106         return pickled_dict
> KeyError: '_hmac'
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ContinuumIO/dask/pull/245#issuecomment-106575694>.
>

cowlicks commented:
Speculating on the problem: When we use our aws credentials we share some connection info across the cluster, when we pass bucket keys around, they have the same connection info so there are no problems. With `anon=True`, each node might be negotiating its own temporary connection with aws, so they have different connection info, so passing bucket keys around breaks things.

I am serializing a connection (a boto bucket object). I connect to s3 on the client, and check the keys exist and find the keys from expanding the pattern. This seems nice to me because each key becomes a partition. But I could close this connection and then make it again for each key, which is probably a lot more traffic but is worth trying.

cowlicks commented:
It would be nice if we could accept `s3://` type URI's  where the URI path gets parsed to a bucket. And also `from_s3(bucket_name, key_name)` with pattern matching of  `*` and `?` in the key name. But when we get `?` and `*` in a URI, we have to do something nontrivial. I think the pattern matching is more useful than accepting `s3://` since the user can just seperate the bucket and key names themselves. So I'm going to punt on taking URI's for now.

cowlicks commented:
I left the s3 URI stuff in since it was almost done, and fixed it with a dirty hack [here](https://github.com/ContinuumIO/dask/pull/245/files#diff-8fd8f45b66540dedbca6875ddd2f3579R811).

I removed the connection serialization. Instead we connect for every key. This fixes the errors with `anon=True`.

Ready for review.

mrocklin commented on dask/bag/core.py:
```diff
@@ -751,6 +758,90 @@ def write(data, filename):
+            connection = boto.connect_s3(anon=anon)
+        elif connection is None:
+            connection = boto.connect_s3(aws_access_key, aws_secret_key)
+        return connection
```
Should this live outside?

mrocklin commented on dask/bag/core.py:
```diff
@@ -751,6 +758,90 @@ def write(data, filename):
+            # read the key
+            read_key = lambda k: [getattr(k, 'read')()]
+            dsk[('load', i)] = (read_key, get_key)
+        return Bag(dsk, 'load', len(matches))
```
Needs a unique name.  Suggest `'load' + next(tokens) or some such.

mrocklin commented on dask/bag/core.py:
```diff
@@ -751,6 +758,90 @@ def write(data, filename):
+            # get a key from the bucket
+            get_key = (lambda b, k: b.get_key(k), get_bucket, m)
+            # read the key
+            read_key = lambda k: [getattr(k, 'read')()]
```
Is this `k.read()`?

mrocklin replied:
These repeated lambdas seem nested and unnecessary.  I might be out of touch though.  What are your thoughts on a top level utility function that takes an s3 uri, creates a bucket, gets a key, then reads the data from the key and returns it.

```python
def _from_s3(bucket_name, key, conn_args):
    conn = boto.connect(...)
    bucket = conn.(...)
    key = bucket.(...)
    return key.read()
```
This has the advantage that we can test this logic in isolation from dask.bag.


If creating connections/buckets turns into a bottleneck then caching might be good.  This would hopefully happen per process.

```python

@toolz.memoize
def boto_connection(conn_args):
    return boto.connect(...)
```

cowlicks replied:
Yeah that is a dumb way of writing ` k.read()` I had some trouble figuring out how to write a dask that does `f()()` or `f().read()`. I naively tried `((f,),)` then this, then `(lambda x: x()(),)`. But calling `key.read` is the wrong thing to do here anyway. I should be doing `list(key)`.

cowlicks replied:
I think what I would want to `memoize` would be a function that connects and gets a bucket. Because we get one bucket per connection, and we get it everytime we make a connection. 

But when I 
```python
@toolz.memoize
def _get_s3_bucket(...):
```
I get serialization errors. So I don't think it `memoize` is working per process...?

mrocklin commented on dask/bag/core.py:
```diff
@@ -751,6 +758,90 @@ def write(data, filename):
+            k_name = k.name.encode('utf-8')
+            if fnmatchcase(k_name, paths):
+                matches.append(k_name)
+        return _from_s3(bucket_name, matches, conn_args)
```
Is there a way to isolate the hacky code into a separate testable function?  This lets us test it more heavily in isolation and isolates the hackiness away from general S3 loading, which is ideally somewhat striaghtforward.

mrocklin commented on dask/bag/tests/test_bag.py:
```diff
@@ -216,6 +216,33 @@ def test_from_filenames_bz2():
+
+    # test with full uri and '?' pattern and auto anon boto connection
+    uri = 's3://nyqpug/tips.c?v'
+    a = db.from_s3(uri, anon=True)
```
Woot.

mrocklin commented on dask/bag/tests/test_bag.py:
```diff
@@ -216,6 +216,33 @@ def test_from_filenames_bz2():
+    assert b.npartitions == 1
+
+    # test all keys in bucket
+    c = db.from_s3('nyqpug')
```
Also woot.

mrocklin replied:
I suspect that this will probably be common case.  I don't know though.  I don't have enough experience using S3 in practice to judge

mrocklin commented:
OK, I would suggest forgetting about memoization for now.  Toolz needs to improve serialization of memoized functions before that can be effective.

mrocklin commented:
We could build caching on our own.  I wouldn't bother though unless you can verify that this is a bottleneck.

mrocklin commented:
https://github.com/pytoolz/toolz/issues/245

cowlicks commented:
@mrocklin I figured it out. I was calling the memoized `_get_s3_bucket` in the main process. Before making the dask, so when each main process called the function they would get it from the main process. And so a bucket (which contains a connection) would get serialized.

The serialization errors were from serializing a connection (not a memoized function) so you can probably close that issue.

I have not tested this with dask distributed yet though.

cowlicks commented on dask/bag/core.py:
```diff
@@ -752,6 +759,103 @@ def write(data, filename):
+    name = next(load_names)
+    dsk = {}
+    for i, k in enumerate(keys):
+        aak, ask, conn, anon = conn_args
```
It would be nice to be able to make a dask like
```python
(func, *args, **kwargs)
```
but this is apparently not valid python.

mrocklin replied:
Yeah, I tend to use `functools.partial` or `toolz.curry` for this.  That bit of syntax we probably can't add.

cowlicks commented on dask/bag/core.py:
```diff
@@ -752,6 +759,103 @@ def write(data, filename):
+        # get a key from the bucket
+        get_key = (lambda b, k: b.get_key(k), get_bucket, k)
+        # read the key
+        dsk[(name, i)] = (list, get_key)
```
Does this look okay?

mrocklin replied:
Thoughts on

```python
def get_key(bucket_name, key_name, **kwargs):
    bucket = _memoized_get_bucket(bucket_name, **kwargs)
    return b.get_key(key_name)
```

Then

```python
def _from_s3(bucket_name, keys, **conn_args):
    name = next(names)
    get_key2 = partial(get_key, **kwargs)
    dsk = dict(((name, i), (get_key2, bucket_name, key)) for i, key in enumerate(keys))
    return Bag(dsk, name, len(keys))
```

This makes the nested lispy code (of which I'm often a culprit) a bit more explicit.  It also lets us test getting a key from s3 in isolation of dask.

mrocklin commented on dask/bag/core.py:
```diff
@@ -752,6 +759,103 @@ def write(data, filename):
+        for k in keys:
+            k_name = k.name
+            if fnmatchcase(k_name, paths):
+                matches.append(k_name)
```
Would this work here?

    matches = [k.name for k in keys if fnmatchcase(k.name, paths)]

mrocklin commented on dask/bag/core.py:
```diff
@@ -752,6 +759,103 @@ def write(data, filename):
+        bucket = _get_s3_bucket(bucket_name, *conn_args)
+
+        # handle globs
+        keys = bucket.list()
```
If possible I generally prefer comments at the end of a line.  My eyes sometimes have trouble walking over very vertical code.

mrocklin commented on dask/bag/tests/test_bag.py:
```diff
@@ -216,6 +217,47 @@ def test_from_filenames_bz2():
+    assert c.npartitions == 3
+
+    d = db.from_s3('s3://nyqpug')
+    assert d.npartitions == 3
```
Can we add a test for gzip and bz2 files?

mrocklin replied:
See the hdfs solution and stream_decompress

cowlicks replied:
Added a gz test. Can I add `tips.bz2` to the `nyqpug` bucket? Where is it at?

mrocklin replied:
cc @cpcloud might know.  Or he might have tips on how to push up data to a new bucket easily (assuming that you have s3 credentials from Troy).  Generally Phil might also have thoughts on this PR and how to make things cleaner.

cowlicks replied:
I can create the new key once I have permission @cpcloud 

for reference
```python
import boto
import bz2
b = boto.connect_s3().get_bucket('nyqpug')
tips = b.get_key('tips.csv').read()
tips_bz2_key = b.new_key('tips.bz2')
tips_bz2_key.set_contents_from_string(bz2.compress(tips))
# permission error here
```

cpcloud commented on dask/bag/core.py:
```diff
@@ -756,6 +764,98 @@ def write(data, filename):
+    if (paths == '*') and (o.path != '' and o.path != '/'):
+        paths = o.path[1:].replace('QUESTIONMARK', '?')
+    bucket_name = o.hostname
+    return bucket_name, paths
```
I have a less hacky way of doing this. @cowlicks Should I submit a PR to your branch?

cowlicks replied:
sure

cowlicks commented:
@mrocklin I created a `tips-data` bucket that continuum gave me access to and mirrored the `nyqpug` data there, and added a `tips.bz2` file.

mrocklin commented:
The from_s3 test are still fairly slow for me.  Is there a config setting I should use for py.test?

mrocklin commented:
This looks good to me.  +1

cowlicks commented:
I had to add a `conftest.py` module to for py.test to know to skip the slow tests.

QCaudron commented:
This looks very useful. However, `dask.bag.from_s3` doesn't seem to be in the docs. Any chance of an update there ? Thanks !

mrocklin commented:
This has been moved to the following:

```python
dask.bag.read_text('s3://bucket/key.*.names')
```

QCaudron commented:
Thanks for the quick reply !

Quentin CAUDRON

On 29 June 2016 at 11:41, Matthew Rocklin <notifications@github.com> wrote:

> This has been moved to the following:
>
> dask.bag.read_text('s3://bucket/key.*.names')
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/dask/dask/pull/245#issuecomment-229449616>, or mute
> the thread
> <https://github.com/notifications/unsubscribe/ADv_G4i7xfp1XNPT7wx1fArRsxIKymzSks5qQrxigaJpZM4ErWeu>
> .
>
